### PR TITLE
V1.0.0-support

### PR DESCRIPTION
**Pull request by bdegeeter**
Update k8s secrets plugin for porter v1.0.0.  
* Remove storage support in favor of new mongodb storage  
* Kubernetes integration testing now done directely in the Porter Operator

### CODE REVIEW

**Comment by carolynvs** [comment]
@bdegeeter When you are ready for a review, just flip this out of draft status. 👍

**Comment by bdegeeter** [comment]
I've got a WIP github actions pipeline.  It's disabled for now and I'll come back to it after addressing any feedback on this PR.

**Review by carolynvs — COMMENTED** [review]
Sorry I haven't finished reviewing this yet. These are some initial suggestions.

File: magefile.go
Comment (carolynvs) [reply magefile.go:148]: This is now at https://github.com/getporter/magesfiles

File: pkg/kubernetes/run.go
Comment (carolynvs) [inline pkg/kubernetes/run.go:49]: I've opened an issue to track making this change back in Porter https://github.com/getporter/porter/issues/1986

File: pkg/kubernetes/secrets/plugin.go
Comment (carolynvs) [inline pkg/kubernetes/secrets/plugin.go:22]: When/how is this used? Is it for running outside of the operator?

File: pkg/kubernetes/secrets/store.go
Comment (carolynvs) [inline pkg/kubernetes/secrets/store.go:52]: What's the significance of pre/post clientset? Is it for just that the kubeconfig's current context had a different namespace set than the store? Or is there more to it?

File: pkg/kubernetes/secrets/store.go
Comment (carolynvs) [inline pkg/kubernetes/secrets/store.go:67]: The connect function returns an error but it's not being checked?

File: README.md
Comment (carolynvs) [inline README.md:66]: Can you add something to the README that explains how this plugin works? It's always been a bit confusing:

* What namespace is used?
* What should the secret name be in k8s?
* What should the secret data keys be? How does that match up with a secret referenced in a credential or parameter set?

File: tests/testdata/test_credentials.yaml
Comment (sgettys) [reply tests/testdata/test_credentials.yaml:2]: This is old, needs to be removed

Comment (sgettys) [reply tests/testdata/test_credentials.yaml:2]: Actually we can leave this here if we want to do local manual testing of the in-operator plugin. It's not currently being used though, I'll update the schema version and just keep it around for now

File: tests/integration/operator/testdata/porter-test-me.yaml
Comment (sgettys) [reply tests/integration/operator/testdata/porter-test-me.yaml:10]: Good point, could we add that to the getporter project as a test bundle? It has a lot of test functionality around inputs, outputs, secrets, and being able to fiddle with the exit code as well as delay
https://github.com/bdegeeter/porter-test-me

File: tests/integration/operator/testdata/Dockerfile.customAgent
Comment (sgettys) [reply tests/integration/operator/testdata/Dockerfile.customAgent:5]: This was because we were always cleaning up our bin before building this so there was only 1 version under /bin/plugins/kubernetes/ but yeah if there's multiple versions under there then this would break. I'll try to clean that up a little bit, not sure if having multiple versions under bin/plugins/kubernetes is a valid use case. Some of this stuff is trying to keep current build process in place and add on top of it without rewriting the whole thing

File: Makefile
Comment (sgettys) [reply Makefile:62]: I'm going to let @bdegeeter take this one, I'm not sure what the thought was for this or if vendoring is strictly required.

File: hack/params.yaml
Comment (sgettys) [reply hack/params.yaml:3]: Ah yeah this is actually being handled via ginkgo now for the integration tests but I'll revert back to using the env vars for the manual test namespace

File: CONTRIBUTING.md
Comment (sgettys) [reply CONTRIBUTING.md:71]: Removed this, the TestLocalIntegration and TestIntegration mage targets are already called out below in the Run a test installation section

File: pkg/kubernetes/secrets/plugin.go
Comment (bdegeeter) [reply pkg/kubernetes/secrets/plugin.go:22]: This needs to go away.  It didn't get cleaned up with some other work.

File: pkg/kubernetes/secrets/store.go
Comment (bdegeeter) [reply pkg/kubernetes/secrets/store.go:52]: No more to it.  It's to help troubleshoot if current env context is different than config

File: Makefile
Comment (bdegeeter) [reply Makefile:62]: I was using a local vendor for a bit to test with some porter code changes.  I'll remove

File: magefile.go
Comment (sgettys) [reply magefile.go:145]: I moved over to what I think is currently implemented here. Might be some more things to add

File: tests/integration/operator/testdata/porter-test-me.yaml
Comment (sgettys) [reply tests/integration/operator/testdata/porter-test-me.yaml:10]: @carolynvs do you want this bundle built and published here or is there an existing repo under getporter I can add this to/create a new one?

File: Makefile
Comment (sgettys) [reply Makefile:5]: I'll move these over to mage

File: tests/integration/operator/testdata/porter-test-me.yaml
Comment (carolynvs) [reply tests/integration/operator/testdata/porter-test-me.yaml:10]: Sorry I missed your original question. It's fine to just use your repository, it doesn't need to be in the porter repository itself.

**Comment by bdegeeter** [comment]
Suggested edit:
```diff
diff --git a/Makefile b/Makefile
index cce4c31..a3042bf 100644
--- a/Makefile
+++ b/Makefile
@@ -30,36 +30,9 @@ else
 FILE_EXT=
 endif
 
-debug: clean build-for-debug  bin/porter$(FILE_EXT)
-
-debug-in-vscode: clean build-for-debug install
-
-build-for-debug:
-	mkdir -p $(BINDIR)
-	$(GO) build -o $(BINDIR)/$(PLUGIN)$(FILE_EXT) ./cmd/$(PLUGIN)
-
-$(BINDIR)/$(PLUGIN)$(FILE_EXT):
-	mkdir -p $(BINDIR)
-	$(GO) build -ldflags '$(LDFLAGS)' -o $(BINDIR)/$(PLUGIN)$(FILE_EXT) ./cmd/$(PLUGIN)
-
-build: $(BINDIR)/$(PLUGIN)$(FILE_EXT)
-
-xbuild-all: test-unit
-	$(foreach OS, $(SUPPORTED_PLATFORMS), \
-		$(foreach ARCH, $(SUPPORTED_ARCHES), \
-				$(MAKE) $(MAKE_OPTS) CLIENT_PLATFORM=$(OS) CLIENT_ARCH=$(ARCH) PLUGIN=$(PLUGIN) xbuild; \
-		))
-
-xbuild: $(BINDIR)/$(VERSION)/$(PLUGIN)-$(CLIENT_PLATFORM)-$(CLIENT_ARCH)$(FILE_EXT)
-$(BINDIR)/$(VERSION)/$(PLUGIN)-$(CLIENT_PLATFORM)-$(CLIENT_ARCH)$(FILE_EXT):
-	mkdir -p $(dir $@)
-	GOOS=$(CLIENT_PLATFORM) GOARCH=$(CLIENT_ARCH) $(XBUILD) -o $@ ./cmd/$(PLUGIN)
-
 test: test-unit test-integration test-in-kubernetes 
 	$(BINDIR)/$(PLUGIN)$(FILE_EXT) version
 
-test-unit: build
-	$(GO) test $(shell go list ./... |grep -v tests/integration );
 test-integration: export CURRENT_CONTEXT=$(shell kubectl config current-context)
 test-integration: bin/porter$(FILE_EXT) setup-tests clean-last-testrun
 	./tests/integration/local/scripts/test-local-integration.sh
@@ -69,9 +42,6 @@ test-integration: bin/porter$(FILE_EXT) setup-tests clean-last-testrun
 		kubectl config use-context $$CURRENT_CONTEXT; \
 	fi
 
-publish: bin/porter$(FILE_EXT)
-	go run mage.go -v Publish $(PLUGIN) $(VERSION) $(PERMALINK)
-
 bin/porter$(FILE_EXT):
 	mkdir -p $(PORTER_HOME)
 	@curl --silent --http1.1 -lfsSLo bin/porter$(FILE_EXT) https://cdn.porter.sh/$(PORTER_VERSION)/porter-$(CLIENT_PLATFORM)-$(CLIENT_ARCH)$(FILE_EXT)
diff --git a/magefile.go b/magefile.go
index 340eb87..2af9985 100644
--- a/magefile.go
+++ b/magefile.go
@@ -9,6 +9,7 @@ import (
 	"log"
 	"os"
 	"path/filepath"
+	"runtime"
 	"strings"
 	"time"
 
@@ -22,6 +23,7 @@ import (
 	"github.com/magefile/mage/mg"
 	"github.com/magefile/mage/target"
 	"github.com/pkg/errors"
+	"golang.org/x/sync/errgroup"
 	// mage:import
 )
 
@@ -35,6 +37,8 @@ const (
 	// Relative location of the KUBECONFIG for the test cluster
 	kubeconfig = "kind.config"
 
+	pluginName = "kubernetes"
+
 	// Namespace of the porter plugin
 	testNamespace = "porter-plugin-test-ns"
 	// Namespace of the porter operator
@@ -53,8 +57,12 @@ const (
 )
 
 // Dirs to watch recursively for build target
-var srcDirs = []string{"cmd/", "pkg/"}
-var binDir = "bin/plugins/kubernetes/"
+var (
+	srcDirs             = []string{"cmd/", "pkg/"}
+	binDir              = "bin/plugins/kubernetes/"
+	pluginPkg           = fmt.Sprintf("./cmd/%s", pluginName)
+	supportedClientGOOS = []string{"linux", "darwin", "windows"}
+)
 
 // Image name for local agent
 var localAgentImgRepository = "localhost:5000/porter-agent-kubernetes"
@@ -66,14 +74,11 @@ var operatorBundleRef = fmt.Sprintf("%s/%s:%s", operatorRegistry, operatorImage,
 // Build a command that stops the build on if the command fails
 var must = shx.CommandBuilder{StopOnError: true}
 
-// We are migrating to mage, but for now keep using make as the main build script interface.
-
 // Publish the cross-compiled binaries.
-// func Publish(plugin string, version string, permalink string) {
-// 	releases.PreparePluginForPublish(plugin, version, permalink)
-// 	releases.PublishPlugin(plugin, version, permalink)
-// 	releases.PublishPluginFeed(plugin, version)
-// }
+func Publish() {
+	PublishPlugin(pluginName)
+	PublishPluginFeed(pluginName)
+}
 
 // Add GOPATH/bin to the path on the GitHub Actions agent
 // TODO: Add to magex
@@ -102,15 +107,46 @@ func Vet() {
 	must.RunV("go", "vet", "./...")
 }
 
+// Run unit tests defined in srcDirs
+func TestUnit() {
+	v := ""
+	if mg.Verbose() {
+		v = "-v"
+	}
+	args := []string{"test", v}
+	for _, dir := range srcDirs {
+		args = append(args, fmt.Sprintf("./%s...", dir))
+	}
+	must.Command("go", args...).CollapseArgs().RunV()
+
+	// Verify integration tests compile
+	must.Run("go", "test", "-run=non", "./tests/...")
+}
+
 func Build() {
 	rebuild, err := target.Dir(binDir, srcDirs...)
 	if err != nil {
-		panic(err)
+		mgx.Must(fmt.Errorf("error inspecting source dirs %s: %w", srcDirs, err))
 	}
 	if rebuild {
 		Clean()
 		mg.Deps(EnsureTestCluster)
-		must.RunV("make", "xbuild-all")
+		TestUnit()
+		//TODO: fix issue with XBuildAll https://github.com/getporter/magefiles/issues/4
+		var g errgroup.Group
+		for _, goos := range supportedClientGOOS {
+			goos := goos
+			g.Go(func() error {
+				return XBuild(pluginPkg, pluginName, binDir, goos, "amd64")
+			})
+		}
+		mgx.Must(g.Wait())
+		info := LoadMetadata()
+		os.RemoveAll(filepath.Join(binDir, "dev"))
+		shx.Copy(filepath.Join(binDir, info.Version), filepath.Join(binDir, "dev"), shx.CopyRecursive)
+		//copy local arch bin to bin/plugins/kubernetes/kubernetes for local integration testing
+		shx.Copy(filepath.Join(binDir, info.Version, fmt.Sprintf("%s-%s-%s", pluginName, runtime.GOOS, "amd64")), filepath.Join(binDir, pluginName))
+		PreparePluginForPublish(pluginName)
 	}
 }
 
@@ -370,14 +406,8 @@ func PublishLocalPorterAgent() {
 
 // Workaround to get the plugin built into an agent image for testing in the operator
 func BuildLocalPorterAgent() {
-	version := "v0"
-	gitMetadata := LoadMetadata()
-	if gitMetadata.Version != "" {
-		version = gitMetadata.Version
-	}
 	buildImage := func(img string) error {
 		_, err := shx.Output("docker", "build", "-t", img,
-			"--build-arg", fmt.Sprintf("PLUGIN_VERSION=%s", version),
 			"--build-arg", fmt.Sprintf("PORTER_VERSION=%s", porterVersion),
 			"--build-arg", fmt.Sprintf("REGISTRY=%s", porterRegistry),
 			"-f", "tests/integration/operator/testdata/Dockerfile.customAgent", ".")
diff --git a/tests/integration/operator/testdata/Dockerfile.customAgent b/tests/integration/operator/testdata/Dockerfile.customAgent
index 5a3ccf9..9312ca9 100644
--- a/tests/integration/operator/testdata/Dockerfile.customAgent
+++ b/tests/integration/operator/testdata/Dockerfile.customAgent
@@ -1,9 +1,4 @@
 ARG PORTER_VERSION
 ARG REGISTRY
-FROM alpine:3.15 as builder
-ARG PLUGIN_VERSION
-COPY --chown=65532:65532 bin/plugins/kubernetes/$PLUGIN_VERSION/kubernetes-linux-amd64 /kubernetes
-
 FROM $REGISTRY/porter-agent:$PORTER_VERSION
-COPY --from=builder --chown=65532:65532 /kubernetes /app/.porter/plugins/kubernetes/kubernetes
-
+COPY --chown=65532:65532 bin/plugins/kubernetes/dev/kubernetes-linux-amd64 /app/.porter/plugins/kubernetes/kubernetes

```

File: magefile.go
Comment (bdegeeter) [reply magefile.go:107]: Moved to issue https://github.com/getporter/kubernetes-plugins/issues/89.

File: CONTRIBUTING.md
Comment (carolynvs) [inline CONTRIBUTING.md:65]: I found that after running this command, I still had to run `docker push` to get the image into a registry accessible by the cluster. Is that right or should I have used a different command?

File: README.md
Comment (carolynvs) [inline README.md:70]: This is what I never understood about the plugin before. Thanks for spelling it out.

But now I'm wondering if we should change the key to something that makes sense for both parameter and credential sets. Possible ideas: `porter-secret`, `value`, `SECRET_NAME` (i.e. if the secret value in the credential set is sadpanda, then the k8s secret is named sadpanda and the key would be sadpanda.

What do you think?

File: mage/docker/docker.go
Comment (carolynvs) [reply mage/docker/docker.go:8]: I'm fine with either waiting and using the magefiles in a separate PR, or doing it here. I am just letting you know in case you missed it that this function can now be used from `[get.porter.sh/magefiles/docker.ExtractRepoDigest](https://github.com/getporter/magefiles/blob/35bfc96f50a849aefe9ba1e1288eb393568bae85/docker/images.go#L10)`

File: mage/tests/kind.config.yaml
Comment (carolynvs) [inline mage/tests/kind.config.yaml:1]: Same comment as above, these test config files aren't needed if you use the targets in the magefiles repo

File: magefile.go
Comment (carolynvs) [inline magefile.go:153]: I did not know about this! 💯

File: magefile.go
Comment (carolynvs) [inline magefile.go:214]: The error from this command isn't being handled. If you want it to fail immediately just replace shx.Command with must.Command. Otherwise you'll want to check the error and handle it.

File: magefile.go
Comment (carolynvs) [inline magefile.go:426]: Is this the function that people should use instead of BuildLocalPorterAgent? I'm referring to the docs in the contributing guide, where we explain which targets to use when.

File: pkg/kubernetes/secrets/store.go
Comment (carolynvs) [reply pkg/kubernetes/secrets/store.go:51]: Let's change this and any others to debug then perhaps? Or do the messages not get captured by porter unless they are info+?

File: pkg/kubernetes/secrets/store.go
Comment (carolynvs) [inline pkg/kubernetes/secrets/store.go:68]: As we were discussing today, the k8s secret plugin should really only resolve value, secret and perhaps in a later PR, config-map. Can you remove this part which falls back to the host plugin? That will remove support for file, env, command, etc.

File: pkg/kubernetes/secrets/store.go
Comment (carolynvs) [inline pkg/kubernetes/secrets/store.go:78]: Let's return an error if the secret doesn't have SecretDataKey. I think that could be a common misunderstanding in how to use the plugin, and any error messages we can surface to help them figure it out will be helpful.

File: tests/integration/operator/ginkgo/installation_test.go
Comment (carolynvs) [inline tests/integration/operator/ginkgo/installation_test.go:250]: Long term it would be great to find a home for these test helpers. I expect over time we will build up more test helper methods and structs that can be shared between any projects working in kubernetes. Perhaps in a test package in operator or something like that.

It's fine to have it copied for now, just putting out the idea of how I'd like to see this evolve going forward so it's easier for us to share stuff like this.

File: tests/integration/operator/ginkgo/suite_test.go
Comment (carolynvs) [inline tests/integration/operator/ginkgo/suite_test.go:18]: nit: Can you do a quick sweep for commented out code and remove lines like this?

File: tests/integration/operator/ginkgo/suite_test.go
Comment (carolynvs) [inline tests/integration/operator/ginkgo/suite_test.go:93]: When we add configmap support to this plugin, I think we'll have to adjust the default porter-agent RBAC to include reading configmaps.

**Review by carolynvs — COMMENTED** [review]
I ran this locally and it worked! I think this is ready to go after a bit of cleanup.

File: mage/docker/docker.go
Comment (sgettys) [reply mage/docker/docker.go:8]: Oh shoot I thought I removed these, I took out all calls to this from the magefile so this can be deleted

File: CONTRIBUTING.md
Comment (sgettys) [reply CONTRIBUTING.md:65]: Yeah there's a separate PushLocalPorterAgent mage command for that.

Comment (sgettys) [reply CONTRIBUTING.md:65]: PushLocalPorterAgent built and pushes to the local registry

File: mage/tests/kind.config.yaml
Comment (sgettys) [reply mage/tests/kind.config.yaml:1]: Removed

**Comment by carolynvs** [comment]
After this is merged, would you be interested in being added as maintainers to this repository? You will know it as well as anyone at that point and I would love the help.

File: tests/integration/operator/ginkgo/suite_test.go
Comment (sgettys) [reply tests/integration/operator/ginkgo/suite_test.go:93]: Think these were the last ones..

File: pkg/kubernetes/secrets/store.go
Comment (sgettys) [reply pkg/kubernetes/secrets/store.go:68]: @carolynvs  We should spell this out in the README then for sure. If the k8s plugin is set then ONLY value, secret (and later config-map) will be supported? Seems like this might be limiting for running this with a local porter instead of the operator is that the behavior we actually want? Maybe I'm not understanding how this would work in that case

File: pkg/kubernetes/secrets/store.go
Comment (sgettys) [reply pkg/kubernetes/secrets/store.go]: @carolynvs added in a custom error for this, just want to double-check wording that it makes sense.

File: tests/integration/operator/ginkgo/suite_test.go
Comment (sgettys) [reply tests/integration/operator/ginkgo/suite_test.go:90]: @carolynvs This will include the RBAC in the operator installer as well right? Or is that already got enough RBAC to grab config-maps?

File: README.md
Comment (sgettys) [reply README.md:70]: Value sounds good to me!

File: pkg/kubernetes/secrets/store.go
Comment (bdegeeter) [reply pkg/kubernetes/secrets/store.go:68]: Does porter fall back to other credSet types if a plugin does not support them?  It's unclear if this behavior change would allow a user to create a credSet with creds of different types.

Comment (carolynvs) [reply pkg/kubernetes/secrets/store.go:68]: I've only ever thought of using this plugin inside kubernetes. I'm not sure why someone would use it for running porter locally but yeah that would cause trouble if we didn't embed the fallback logic directly into the plugin since Porter doesn't try other plugins if the first one can't resolve it. There's only one secret plugin active at a time.

It doesn't hurt anything to leave the host plugin in there.

File: tests/integration/operator/ginkgo/suite_test.go
Comment (carolynvs) [reply tests/integration/operator/ginkgo/suite_test.go:90]: Right now the agent role has access to secrets but not config maps. So when we add support for config map the operator rbac will need to be updated to grant that as well to the porter-agent role.

File: pkg/kubernetes/secrets/store.go
Comment (carolynvs) [inline pkg/kubernetes/secrets/store.go:79]: I think it won't be clear enough what to do with this error message

Key credential not found in secret

How about we  provide a bit more context and what they should do to fix it?

The secret NAMESPACE/SECRET_NAME does not have a key named SECRETDATAKEY. The kubernetes.secrets plugin requires that the Kubernetes secret is named after the secret referenced in the Porter parameter or credential set, and secret value is stored in a key on the Kubernetes secret named SECRETDATAKEY.

Though I still think we need to also change the name of the secret data key since credential is a bit misleading when we are using the plugin to resolve parameter set secrets too.

**Review by carolynvs — APPROVED** [review]
:shipit: This is great, thank you so much for redoing the plugin to work with the porter v1 and the latest operator!

**Comment by carolynvs** [comment]
Oops, can you resolve the conflict on go.mod/go.sum and then I'll merge?

**Comment by sgettys** [comment]
@carolynvs resolved conflicts

**Pull request merged**
